### PR TITLE
Adding basic app infrastructure and models

### PR DESCRIPTION
**Pull request by awais786**
ECOM-4892
Adding new app authoring.
Adding models for authoring app.
Update existing models for about course requirements.
Added a new page to view all non-published course runs. ( page is without any design/css)

### CODE REVIEW

File: course_discovery/apps/authoring/models.py [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+from django_extensions.db.models import TimeStampedModel
+from simple_history.models import HistoricalRecords
+
+from course_discovery.apps.core.models import User
+from course_discovery.apps.course_metadata.models import CourseRun, Program
+
+logger = logging.getLogger(__name__)
+
+
+class KeyWord(TimeStampedModel):
```
Comment (clintonb) [inline course_discovery/apps/authoring/models.py]: Keyword

File: course_discovery/apps/authoring/models.py [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+    """ Keywords model. """
+    name = models.CharField(max_length=255)
+
+    def __str__(self):
+        return self.name
+
+
+class WorkFlowCourse(TimeStampedModel):
```
Comment (clintonb) [inline course_discovery/apps/authoring/models.py]: Workflow

**Comment by awais786** [comment]
@clintonb Thanks! I have made the suggested changes. But this is still under progress and not ready for review.

**Comment by awais786** [comment]
@bderusha Please share your thoughts about the these models.

File: course_discovery/apps/course_metadata/models.py [hunk 3]
```diff
@@ -53,6 +53,30 @@ class Meta(object):
         abstract = True
 
 
+class AbstractSocialNetworkModel(TimeStampedModel):
+    """ SocialNetwork model. """
+    FACEBOOK = 'facebook'
+    TWITTER = 'twitter'
+    Blog = 'blog'
```
Comment (bderusha) [inline course_discovery/apps/course_metadata/models.py]: `BLOG`

File: course_discovery/apps/course_metadata/models.py [hunk 4]
```diff
@@ -53,6 +53,30 @@ class Meta(object):
...
+    OTHERS = 'others'
+
+    SOCIAL_NETWORK_CHOICES = (
+        (FACEBOOK, _('Facebook')),
+        (TWITTER, _('Twitter')),
+        (Blog, _('Blog')),
```
Comment (bderusha) [inline course_discovery/apps/course_metadata/models.py]: `BLOG`

File: course_discovery/apps/authoring/models.py [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+
+
+class WorkflowDetail(TimeStampedModel):
+    """ WorkflowDetail model. """
+
+    workflow = models.OneToOneField(Workflow, related_name='work_flow_details')
+    is_re_run = models.BooleanField(null=False, blank=False, default=False)
+    school = models.CharField(max_length=255, null=True, blank=True)
+    target_content = models.BooleanField(default=False)
+    instructors = SortedManyToManyField(Person, blank=True, related_name='workflow_instructed')
```
Comment (awais786) [inline course_discovery/apps/authoring/models.py]: These fields (instructors, staff)  already exists into course_run. I think we can avoid these here.

**Comment by awais786** [comment]
We have some more fields in https://openedx.atlassian.net/browse/ECOM-4886. Should we add those fields in table now ?

```
Date Submitted
Target Content (Y/N)
Priority
Date Updated
School/Institution
Course name
Date moved to Current Status
Status
```

File: course_discovery/apps/authoring/tests/__init__.py [hunk 6]
```diff
@@ -0,0 +1,21 @@
+from waffle.models import Switch
+
+
+def toggle_switch(name, active):
```
Comment (bderusha) [inline course_discovery/apps/authoring/tests/__init__.py]: Why is this in test?

Comment (awais786) [inline course_discovery/apps/authoring/tests/__init__.py]: Helper method for toggling the status for testing purpose.

File: course_discovery/settings/base.py [hunk 7]
```diff
@@ -49,10 +49,10 @@
     'course_discovery.apps.catalogs',
     'course_discovery.apps.course_metadata',
     'course_discovery.apps.edx_haystack_extensions',
+    'course_discovery.apps.authoring',
```
Comment (bderusha) [inline course_discovery/settings/base.py]: Sorry to be a pain.  But I think that the official name for this app based on how we expect it to be used long term is `publisher`.  Please make the appropriate changes throughout

File: course_discovery/templates/course_builder/course_listing.html [hunk 8]
```diff
@@ -0,0 +1,9 @@
+{% extends 'base.html' %}
```
Comment (bderusha) [inline course_discovery/templates/course_builder/course_listing.html]: Please rename this file path.

File: course_discovery/settings/base.py (hunk 7 above)
Comment (awais786) [inline course_discovery/settings/base.py]: App renamed.

**Comment by awais786** [comment]
@waheedahmed Please review this PR. And you can implement the listing page in this PR also.

**Comment by awais786** [comment]
Listing page [Ticket](https://openedx.atlassian.net/browse/ECOM-4886).

**Comment by awais786** [comment]
@bderusha you can review this PR again.

File: course_discovery/apps/course_metadata/models.py [hunk 9]
```diff
@@ -55,6 +57,30 @@ class Meta(object):
...
+        (TWITTER, _('Twitter')),
+        (BLOG, _('Blog')),
+        (OTHERS, _('Others')),
+    )
+
+    type = models.CharField(max_length=15, choices=SOCIAL_NETWORK_CHOICES, db_index=True)
+    value = models.CharField(max_length=500)
+
+    def __str__(self):
+        return '{type}: {value}'.format(type=self.type, value=self.value)
```
Comment (bderusha) [inline course_discovery/apps/course_metadata/models.py:78]: add a test for this please

File: course_discovery/apps/course_metadata/models.py [hunk 10]
```diff
@@ -64,6 +90,7 @@ class Image(AbstractMediaModel):
 class Video(AbstractMediaModel):
     """ Video model. """
     image = models.ForeignKey(Image, null=True, blank=True)
+    languages = models.ManyToManyField(LanguageTag, null=True, blank=True, related_name='videos')
```
Comment (bderusha) [inline course_discovery/apps/course_metadata/models.py]: `transcript_languages`

File: course_discovery/apps/course_metadata/models.py [hunk 11]
```diff
@@ -137,6 +178,19 @@ class Course(TimeStampedModel):
...
+    course_number = models.CharField(
+        max_length=10, validators=[
+            RegexValidator(
+                regex=COURSE_NUMBER_REGEX,
+                message="Alphanumeric characters and dot (.) are allowed. With lower (x) at end.",
+                code="invalid_course_number",
+            )
+        ],
+        null=True, blank=True, help_text=_(
+            "Course number denoted by .1, .2, etc. at the end of the course number before the `x`"))
```
Comment (bderusha) [inline course_discovery/apps/course_metadata/models.py]: This help text is confusing.  Is there a better way to write this?

File: course_discovery/apps/publisher/tests/test_views.py [hunk 12]
```diff
@@ -0,0 +1,31 @@
+"""
+Tests for course builder views.
```
Comment (bderusha) [inline course_discovery/apps/publisher/tests/test_views.py]: update doc string

File: course_discovery/apps/publisher/tests/test_views.py [hunk 13]
```diff
@@ -0,0 +1,31 @@
...
+"""
+from django.core.urlresolvers import reverse
+from django.test import TestCase
+
+from course_discovery.apps.publisher.tests import toggle_switch
+
+
+class CourseListing(TestCase):
+    """ Course builder list view tests. """
```
Comment (bderusha) [inline course_discovery/apps/publisher/tests/test_views.py]: update doc string

File: course_discovery/apps/publisher/urls.py [hunk 14]
```diff
@@ -0,0 +1,9 @@
+"""
+URLs for the course builder views.
```
Comment (bderusha) [inline course_discovery/apps/publisher/urls.py]: update doc string

File: course_discovery/apps/publisher/views.py [hunk 15]
```diff
@@ -0,0 +1,33 @@
+"""
+Course Builder views.
```
Comment (bderusha) [inline course_discovery/apps/publisher/views.py]: update doc string

File: course_discovery/apps/publisher/models.py [hunk 16]
```diff
@@ -0,0 +1,91 @@
...
+    course_length = models.PositiveIntegerField(null=True, blank=True, help_text=_(
+        "Length of course, in number of weeks"))
+
+    target_content = models.BooleanField(default=False)
+    priority = models.BooleanField(default=False)
+
+    history = HistoricalRecords()
+
+    def __str__(self):
+        return '{course_run}: {program}'.format(course_run=self.course_run, program=self.program)
```
Comment (bderusha) [inline course_discovery/apps/publisher/models.py]: add test

File: course_discovery/apps/publisher/models.py [hunk 17]
```diff
@@ -0,0 +1,91 @@
...
+    name = models.CharField(max_length=15, choices=STATUS_CHOICES, db_index=True)
+
+    updated_by = models.ForeignKey(User, related_name='status_updated_by')
+    history = HistoricalRecords()
+
+    def __str__(self):
+        return '{course_run}'.format(course_run=self.course_run.key)
+
+
+class WorkflowProgram(TimeStampedModel):
```
Comment (bderusha) [inline course_discovery/apps/publisher/models.py]: We need a consistent naming convention.  Why is this one `WorkflowProgram`and below `CourseRunDetail`?

File: course_discovery/apps/publisher/models.py [hunk 18]
```diff
@@ -0,0 +1,91 @@
...
+
+    def __str__(self):
+        return '{type}: {name}'.format(type=self.type, name=self.name)
+
+
+class CourseRunDetail(TimeStampedModel):
+    """ CourseRunDetail model. It contains fields related with
+    course-run."""
+
+    XSERIES = 'xseries'
```
Comment (bderusha) [inline course_discovery/apps/publisher/models.py:43]: these constants are unused

File: course_discovery/apps/publisher/models.py [hunk 19]
```diff
@@ -0,0 +1,91 @@
...
+
+    PROGRAMS_CHOICES = (
+        (XSERIES, _('XSeries')),
+        (MICRO_MASTERS, _('Micro-Masters')),
+    )
+
+    course_run = models.OneToOneField(CourseRun, related_name='detail')
+    is_re_run = models.BooleanField(default=True)
+
+    program = models.ForeignKey(WorkflowProgram, related_name='program_work_flow')
```
Comment (bderusha) [inline course_discovery/apps/publisher/models.py]: I want to know more about why we are storing this information.  We may be able to get away with just listing choices without creating a new table

File: course_discovery/apps/publisher/urls.py [hunk 20]
```diff
@@ -0,0 +1,9 @@
...
+"""
+from django.conf.urls import url
+from course_discovery.apps.publisher.views import CourseListing
+
+urlpatterns = [
+    url(r'^unpublished_courses/$', CourseListing.as_view(), name='unpublished_courses'),
```
Comment (bderusha) [inline course_discovery/apps/publisher/urls.py]: url can be `courses`

**Comment by bderusha** [comment]
Review pass complete.  Please add tests for those `__str__` methods and update the url/paths to not be specific to unpublished courses.

File: course_discovery/apps/publisher/models.py (hunk 19 above)
Comment (awais786) [inline course_discovery/apps/publisher/models.py]: Moving these fields into detail table.

File: course_discovery/templates/publisher/unpublished_courses.html [hunk 21]
```diff
@@ -0,0 +1,25 @@
+{% extends 'base.html' %}
+{% load i18n %}
+{% block title %}
+    {% trans "Course Listing Page" %}
+{% endblock title %}
+
+{% block content %}
+    Course Listing Page
+    <br>
+    {% for course_run in courses_runs %}
```
Comment (bderusha) [inline course_discovery/templates/publisher/unpublished_courses.html]: Can we put these into a table.  That is ultimately how this data will need to be displayed.

Comment (awais786) [inline course_discovery/templates/publisher/unpublished_courses.html]: Sure i will do this.

File: course_discovery/apps/publisher/models.py (hunk 18 above)
Comment (awais786) [inline course_discovery/apps/publisher/models.py:43]: I am using them in choices tuple.

**Pull request closed without merging**